test/gtest/type-builder.cpp

```diff
@@ -533,3 +533,35 @@ TEST_F(NominalTest, TestSubTypes) {
   auto subTypes1 = subTypes.getStrictSubTypes(built[1]);
   EXPECT_EQ(subTypes1.size(), 0u);
 }
+
+// Test reuse of a previously built type as supertype.
+TEST_F(NominalTest, TestExistingSubType) {
+  // Build an initial type A
+  Type A;
+  {
+    TypeBuilder builder(1);
+    builder[0] = Struct();
+    auto result = builder.build();
+    ASSERT_TRUE(result);
+    auto built = *result;
+    A = Type(built[0], Nullable);
+  }
+
+  // Build a type B <: A using a new builder
+  Type B;
+  {
+    TypeBuilder builder(1);
+    builder[0] = Struct();
+    builder.setSubType(0, A.getHeapType());
+    auto result = builder.build();
+    ASSERT_TRUE(result);
+    auto built = *result;
+    B = Type(built[0], Nullable);
+  }
+
+  // Test that B <: A where A is the initial type A
+  auto superOfB = B.getHeapType().getSuperType();
+  ASSERT_TRUE(superOfB);
+  EXPECT_EQ(*superOfB, A.getHeapType());
+  EXPECT_NE(B.getHeapType(), A.getHeapType());
+}
```

Comment:
Could you add one more test in isorecursive mode checking that if we build `A`, then separately we build `A'`, `B <: A`, and `B' <: A'`, then at the end we still have `A == A'` and `B == B'`? That will test that there is no difference between existing and new supertypes as far as isorecursive canonicalization goes.

Reply:
Currently looking into the additional test. So far, when I make a type `A1` using its own builder, an identical type `A2` using its own builder, and then a type `B1 <: A1` using its own builder, building fails with `"Heap type has an undeclared supertype" at index 0`. So I guess there's something else missing. Ideas?

Reply:
Ah right, good find. This code needs to be updated to differentiate between canonical and temporary supertypes that haven't been seen yet. The canonical types should be allowed, but the temporary types should continue producing this error. There's an `isTemp(HeapType)` function you can use to check this.